pkg/infrastructure/azure/azure.go

```diff
@@ -26,3 +45,282 @@ func InitializeProvider() clusterapi.Provider {
 func (p *InfraProvider) Name() string {
 	return "azure"
 }
+
+// PreProvision is called before provisioning using CAPI controllers has begun.
+func (p *InfraProvider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInput) error {
+	session, err := in.InstallConfig.Azure.Session()
+	if err != nil {
+		return fmt.Errorf("failed to get session: %w", err)
+	}
+	// sessionClient := icazure.NewClient(session)
+
+	installConfig := in.InstallConfig.Config
+	platform := installConfig.Platform.Azure
+	subscriptionID := session.Credentials.SubscriptionID
+
+	var cloudConfiguration cloud.Configuration
+	switch platform.CloudName {
+	case aztypes.PublicCloud:
+		cloudConfiguration = cloud.AzurePublic
+	case aztypes.USGovernmentCloud:
+		cloudConfiguration = cloud.AzureGovernment
+	}
+
+	resourceGroupName := platform.ClusterResourceGroupName(in.InfraID)
+	storageAccountName := fmt.Sprintf("cluster%s", randomStringPrefixFunction())
+	containerName := "vhd"
+	blobName := fmt.Sprintf("rhcos%s.vhd", randomStringPrefixFunction())
+
+	stream, err := rhcos.FetchCoreOSBuild(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get rhcos stream: %w", err)
+	}
+	archName := arch.RpmArch(string(installConfig.ControlPlane.Architecture))
+	streamArch, err := stream.GetArchitecture(archName)
+	if err != nil {
+		return fmt.Errorf("failed to get rhcos architecture: %w", err)
+	}
+
+	azureDisk := streamArch.RHELCoreOSExtensions.AzureDisk
+	imageURL := azureDisk.URL
+	// rawImageVersion := strings.ReplaceAll(azureDisk.Release, "-", "_")
+	// imageVersion := rawImageVersion[:len(rawImageVersion)-6]
+
+	headResponse, err := http.Head(imageURL)
+	if err != nil {
+		return fmt.Errorf("failed HEAD request for image URL %s: %w", imageURL, err)
+	}
+
+	imageLength := headResponse.ContentLength
+	if imageLength%512 != 0 {
+		return fmt.Errorf("image length is not alisnged on a 512 byte boundary")
+	}
+
+	userTags := platform.UserTags
+	tags := make(map[string]*string, len(userTags)+1)
+	tags[fmt.Sprintf("kubernetes.io_cluster.%s", in.InfraID)] = ptr.To("owned")
+	for k, v := range userTags {
+		tags[k] = ptr.To(v)
+	}
+
+	tokenCredential := session.TokenCreds
+
+	// Create resource group
+	resourcesClientFactory, err := armresources.NewClientFactory(
+		subscriptionID,
+		tokenCredential,
+		&arm.ClientOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloudConfiguration,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get azure resource groups factory: %w", err)
+	}
+	resourceGroupsClient := resourcesClientFactory.NewResourceGroupsClient()
+	resourceGroup, err := resourceGroupsClient.CreateOrUpdate(
+		ctx,
+		resourceGroupName,
+		armresources.ResourceGroup{
+			Location:  ptr.To(platform.Region),
+			ManagedBy: nil,
+			Tags:      tags,
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("error creating resource group %s: %w", resourceGroupName, err)
+	}
+	logrus.Debugf("ResourceGroup.ID=%s", *resourceGroup.ID)
```

Comment:
> to the InfraReady hook and

After further review, it's probably better to move it to the Ignition hook. 

We will need to upload ignition to the storage account & bucket you're creating here. To do that we will need the storage account name, which right now includes a random prefix. To solve all this it may be easiest to just move this to the ignition hook (you probably need to expand the input parameters to the ignition hook but feel free).